instrumentation/reactor-3.1/library/src/main/java/io/opentelemetry/instrumentation/reactor/ReactorTracing.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.reactor;
+
+import io.opentelemetry.context.Context;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class ReactorTracing {
+
+  private static final Object TRACE_CONTEXT_KEY =
+      new Object() {
+        @Override
+        public String toString() {
+          return "otel-trace-context";
+        }
+      };
+
+  static volatile Mono<String> dummyMono;
+  static volatile Flux<String> dummyFlux;
+
+  static synchronized void enable() {
+    // have to be reset as they capture async strategy and lift
+    dummyMono = Mono.just("");
+    dummyFlux = Flux.just("");
+  }
+
+  static synchronized void disable() {
+    // have to be reset as they capture async strategy and lift
+    dummyMono = Mono.just("");
+    dummyFlux = Flux.just("");
+  }
+
+  /** Forces Mono to run in traceContext scope. */
+  static <T> Mono<T> runInScope(Mono<T> publisher, Context tracingContext) {
+    // this hack forces 'publisher' to run in the onNext callback of `TracingSubscriber`
+    // created for this publisher and with current() span that refer to span created here
+    // without the hack, publisher runs in the onAssembly stage, before traceContext is made current
+    return dummyMono
+        .flatMap(i -> publisher)
+        .subscriberContext(ctx -> storeOpenTelemetryContext(ctx, tracingContext));
+  }
+
+  /** Forces Flux to run in traceContext scope. */
+  static <T> Flux<T> runInScope(Flux<T> publisher, Context tracingContext) {
+    // this hack forces 'publisher' to run in the onNext callback of `TracingSubscriber`
+    // created for this publisher and with current() span that refer to span created here
+    // without the hack, publisher runs in the onAssembly stage, before traceContext is made current
+    return dummyFlux
+        .flatMap(i -> publisher)
+        .subscriberContext(ctx -> storeOpenTelemetryContext(ctx, tracingContext));
+  }
+
+  /**
+   * Stores Trace {@link io.opentelemetry.context.Context} in Reactor {@link
+   * reactor.util.context.Context}.
+   *
+   * @param context Reactor's context to store trace context in.
+   * @param traceContext Trace context to be stored.
+   */
+  public static reactor.util.context.Context storeOpenTelemetryContext(
```

Comment:
Instead of exposing `store/get`, can we expose `runInScope` (`runWithContext`)? It seems to be the better API

Reply:
We can, but it implies using a hack with dummy mono/flux even when it's not needed (not a big deal). But I think we ultimately don't need `runWithContext` - having something like `tracePublisher(Publisher, SpanBuilder, AttributeExtractors?, whatnot)`  will eliminate the need for it and provide convenience.

For those who want low-level control (e.g. if I want to start/end a span conditionally or in some funky way), get/set context is the minimal needed functionality

Reply:
also, if I need to inject headers or enrich span, I'd prefer explicit context from reactor (i.e. `get` method) over `Context.current()`. Since current can change in any user callback, I never know what it refers to. There are no guarantees with explicit reactor context either, so it's rather a preference.